iOS/macOS snapshotter

ivovandongen opened the pull request:
iOS/macOS companion for #9748 

TODO:
- [x] Work out options class
- [x] Extend test view
- [x] Documentation

frederoni commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+ @param mapCamera the camera settings
+ @param size the image size
+ */
+- (instancetype )initWithStyleURL:(NSURL* )styleURL mapCamera:(MGLMapCamera*) mapCamera size:(CGSize) size;
```
Nit: Inconsistent formatting. (only spaces before `*`)
Drop `map` from `mapCamera`. (×3)

frederoni commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+@property (nonatomic) MGLMapCamera* mapCamera;
+
+/**
+ (Optionally) a region to capture. Overrides the center coordinate
```
Let's drop optional since region is invalid by default and Obj-C doesn't support optional structs.

frederoni commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+/**
+ A block to processes the result or error of a snapshot request.
+ 
+ The result will be either an `MGLImage*` or a `NSError*`
```
Nit: s/MGLImage*/MGLImage (I don't think jazzy will link otherwise)

frederoni commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+- (void)cancel;
+
+/**
+ Indicates wether as snapshot is currently being generated.
```
typo: s/wether/whether

frederoni commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,120 @@
+    [self startWithQueue:dispatch_get_main_queue() completionHandler:completion];
+}
+
+- (void)startWithQueue:(dispatch_queue_t)queue completionHandler: (MGLMapSnapshotCompletionHandler)completion;
```
This is slightly confusing. `startWithQueue` implies that we generate the snapshot on the given queue but here we're delegating back on the given queue. This is usually called `handlerQueue` or `delegateQueue`.
I think we should dispatch to the given queue before handing off to core so we can decide if we want to snapshot in parallel or serial.

frederoni commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,120 @@
+    std::unique_ptr<mbgl::MapSnapshotter> mbglMapSnapshotter;
+    std::unique_ptr<mbgl::Actor<mbgl::MapSnapshotter::Callback>> snapshotCallback;
+    
+    BOOL loading;
```
Nit: loading is already synthesized. This can be dropped. Use `_loading` to ignore the accessor when you set the variable.

frederoni commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,120 @@
+    mbglMapSnapshotter.reset();
+}
+
+-(BOOL)isLoading;
```
Nit: already synthesized. Drop this method.

ivovandongen commented:
@frederoni Thanks for the review! I've fixed up the comments and made sure the work is now all done on the provided queue (or the main queue).

I also added some code to add the logo on the image in a work queue (https://github.com/mapbox/mapbox-gl-native/pull/9891/files#diff-56bc5c1e41273957efd873aa83e6c522R111). Could you check if that is the right queue to do this on?

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+ */
+- (instancetype)initWithStyleURL:(NSURL*)styleURL camera:(MGLMapCamera*)camera size:(CGSize)size;
+
+#pragma mark - Configuring the map
```
For our public headers, we use Title Case in marks, since jazzy uses the marks to produce section headings in the generated documentation. We could move to sentence case for that, but it would be nice to do it consistently across all the headers when we make that decision.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+/**
+  The style URL for these options.
+ */
+@property (nonatomic, readonly) NSURL* styleURL;
```
FYI, the predominant style in our Objective-C code is to place the space before the asterisk. I realize this is different than the predominant style for our C++ code.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+/**
+ The zoom. Default is 0.
+ */
+@property (nonatomic) double zoom;
```
The iOS/macOS APIs consistently use the term “zoom level” rather than “zoom” for this concept. The main reason is that “zoom” is also a verb; in Objective-C, `[mapView zoom]` looks like a call to an action method rather than a reference to a property getter.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+ A region to capture. Overrides the center coordinate
+ in the mapCamera options if set
+ */
+@property (nonatomic) MGLCoordinateBounds region;
```
Most of the API uses the term “coordinate bounds” instead of “region”, although there is some holdover from MapKit in the delegate methods. It’s a confusing situation overall, but I think it would be slightly less confusing to use `coordinateBounds` here, to align with `MGLMapView.visibleCoordinateBounds`.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+#pragma mark - Configuring the image
+
+/**
+ The size of the output image. Minimum is 64x64
```
Is this size measured in points or pixels?

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+/**
+ A block to processes the result or error of a snapshot request.
+ 
+ The result will be either an `MGLImage` or a `NSError`
```
`MGLImage` is an unfamiliar typedef that we don’t document anywhere. Instead, create duplicate declarations, one using NSImage and the other using UIImage, and have the compiler switch between them using conditional compilation. See MGLStyle.h for an example.

fabian-guerra replied:
This code is similar to `MGLStyle.h` https://github.com/mapbox/mapbox-gl-native/blob/master/platform/darwin/src/MGLStyle.h#L523 is this ok?

1ec5 replied:
There’s nothing technically wrong with using `MGLImage` as the return value, since the compiler (and thus jazzy) will automatically substitute NSImage or UIImage as appropriate. But neither Quick Help nor jazzy will replace `MGLImage` inside a documentation comment. The same goes for `MGLColor`; that’s why we [use conditional compilation](https://github.com/mapbox/mapbox-gl-native/blob/53d072cbf539b30301fe3b77336751bf2c29af89/platform/darwin/src/MGLLineStyleLayer.h#L243-L299) whenever we need to refer specifically to the underlying type in documentation.

Besides, this sentence isn’t accurate: the block doesn’t return anything (`void`), so there’s no result to speak of. Rather, I think the intention was to say, “Either `snapshot` or `error` will be non-nil,” but that’s already suitably captured in the following two `@param` lines. So I think this line can simply be removed.

fabian-guerra replied:
What I meant is since `MGLStyle.h` is not using conditional compilation I should go ahead and add it there too.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+- (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshotCompletionHandler)completionHandler;
+
+/**
+ Cancels the snapshot creation request, if any.
```
Can this method be called more than once on the same snapshotter object?

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,163 @@
+            if (mbglError) {
+                NSString *description = @(mbgl::util::toString(mbglError).c_str());
+                NSDictionary *userInfo = @{NSLocalizedDescriptionKey: description};
+                NSError *error = [NSError errorWithDomain:MGLErrorDomain code:1 userInfo:userInfo];
```
This arbitrary code should be replaced with a new value of MGLErrorCode defined in MGLTypes.h.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,163 @@
+                    NSImage *compositedImage = mglImage;
+                    
+                    [compositedImage lockFocus];
+                    [logoImage drawInRect:CGRectMake(8, 8, logoImage.size.width,logoImage.size.height)];
```
For maintainability, we should move these eights into a constant CGPoint.

1ec5 commented on platform/ios/app/MBXSnapshotsViewController.m:
```diff
@@ -0,0 +1,66 @@
+    // Top row
+    MGLMapSnapshotter* snapshotterTL;
+    MGLMapSnapshotter* snapshotterTM;
+    MGLMapSnapshotter* snapshotterTR;
```
FYI, the convention in Objective-C is to use a naming scheme like `topCenterSnapshotter`.

1ec5 commented on platform/ios/app/MBXSnapshotsViewController.m:
```diff
@@ -0,0 +1,66 @@
+    MGLMapCamera* mapCamera = [[MGLMapCamera alloc] init];
+    mapCamera.pitch = 20;
+    mapCamera.centerCoordinate = coordinates;
+    MGLMapSnapshotOptions* options = [[MGLMapSnapshotOptions alloc] initWithStyleURL:[NSURL URLWithString:@"mapbox://styles/mapbox/traffic-day-v2"] camera:mapCamera size:CGSizeMake(imageView.frame.size.width, imageView.frame.size.height)];
```
MGLStyle has factory methods for style URLs. We use them whenever possible.

fabian-guerra replied:
In this case the traffic style is deprecated, even tho it has not yet merged into master. Our recommendation is create an URL with this string `mapbox://styles/mapbox/traffic-day-v2`.

1ec5 replied:
In this case, the demo doesn’t depend on the traffic congestion information displayed by the Traffic Day style, so I’d recommend choosing a non-deprecated method like `+[MGLStyle satelliteStreets]`.

1ec5 commented on platform/macos/app/MapDocument.m:
```diff
@@ -153,6 +157,67 @@ - (NSURL *)shareURL {
+                        
+                    }
+                    
+                    NSString *extension = [[fileURL pathExtension] lowercaseString];
```
File name extensions aren’t especially reliable on Apple platforms, since a file’s canonical type is represented by a [UTI](https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html). The save panel even lets you save without an extension unless you restrict the allowed file types to those that can be represented as images:

```objc
panel.allowedFileTypes = [NSImage imageTypes];
```

Here’s [a more robust way](https://stackoverflow.com/a/1506350/4585461) to go from a file extension to a UTI.

1ec5 replied:
An alternative to looking at file extensions:

1. The save panel tells us the file path. use NSFileManager to [determine the UTI at that path](https://stackoverflow.com/questions/25498417/how-do-i-get-a-document-uti-from-a-file-path#comment39847300_25498566). (Not sure if this only works for existing files or if it would work for unsaved files as well.)
1. Use `UTTypeConformsTo()` to determine whether the UTI at that path conforms to `public.jpeg` etc.
1. Avoid hard-coding UTIs like `public.jpeg` by iterating over `+[NSImage imageTypes]`.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+ @param camera the camera settings
+ @param size the image size
+ */
+- (instancetype)initWithStyleURL:(NSURL*)styleURL camera:(MGLMapCamera*)camera size:(CGSize)size;
```
Do we want to allow the style URL to be `nil` and default to Mapbox Streets, as MGLMapView and MGLOfflineStorage do?

fabian-guerra replied:
@1ec5 could you please help me understand why defaulting to Mapbox streets is a good option.

1ec5 replied:
When you create an MGLMapView or MGLOfflineRegion with a `nil` style URL, we default to the current version of Mapbox Streets:

https://github.com/mapbox/mapbox-gl-native/blob/53d072cbf539b30301fe3b77336751bf2c29af89/platform/ios/src/MGLMapView.mm#L397
https://github.com/mapbox/mapbox-gl-native/blob/53d072cbf539b30301fe3b77336751bf2c29af89/platform/darwin/src/MGLTilePyramidOfflineRegion.mm#L36

For consistency with these APIs, I think we should treat a `nil` style URL here the same way.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.h:
```diff
@@ -0,0 +1,105 @@
+
+/**
+ Creates a set of options with the minimum required information
+ @param styleURL the style url to use
```
Nit: Start with a capital letter and end with a period. This makes things more consistent in the event that we have to cram more than one sentence in a parameter description, which isn’t uncommon.

1ec5 commented on platform/darwin/src/MGLMapSnapshotter.mm:
```diff
@@ -0,0 +1,163 @@
+- (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshotCompletionHandler)completion;
+{
+    if ([self isLoading]) {
+        NSDictionary *userInfo = @{NSLocalizedDescriptionKey: @"Already started this snapshotter"};
```
This string should be [localizable](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/DEVELOPING.md#adding-user-facing-text), since it can be presented to the user.

boundsj commented:
> update the changelog to mention the new feature

For iOS: https://github.com/mapbox/mapbox-gl-native/pull/9919

1ec5 commented:
Filed #9920 so we can better track the fit-and-finish detailed in https://github.com/mapbox/mapbox-gl-native/pull/9891#pullrequestreview-60754801.

1ec5 commented on platform/macos/app/Base.lproj/MainMenu.xib:
```diff
@@ -128,6 +128,12 @@
                                     <action selector="revertDocumentToSaved:" target="-1" id="iJ3-Pv-kwq"/>
                                 </connections>
                             </menuItem>
+                            <menuItem title="Save snapshot" id="vjX-0E-kLO">
```
Use title case on menu items.
